presto-common/src/main/java/com/facebook/presto/common/type/TypeSignatureUtils.java

```diff
@@ -0,0 +1,158 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.common.type;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import static java.util.Collections.unmodifiableList;
+
+public final class TypeSignatureUtils
+{
+    private TypeSignatureUtils() {}
+
+    public static TypeSignature resolveIntermediateType(TypeSignature actualTypeSignature, List<TypeSignature> actualTypeSignatureParameters, List<TypeSignature> expectedTypeSignatureParameters)
```

Comment:
why do we need special logic to handle aggregation function intermediate types?

Reply:
We only need special logic to handle generic aggregate functions. 
To resolve the intermediate type for generic aggregate functions, we rely on utility functions from TypeSignatureUtils. Specifically, we use the parameter types passed down in the function signatures to determine and insert the appropriate intermediateType.

- Why do we need this?
This is required because aggregate function metadata for generic aggregate functions contain generic intermediate args i.e `K,V` and we need a way to resolve these args.

Example:
`max` function:
For the max function, here is the relevant metadata:

```
{
  "aggregateMetadata": {
    "intermediateType": "t",
    "isOrderSensitive": false
  },
  "docString": "native.default.max",
  "functionKind": "AGGREGATE",
  "outputType": "t",
  "paramTypes": [
    "t"
  ],
  "routineCharacteristics": {
    "determinism": "DETERMINISTIC",
    "language": "CPP",
    "nullCallClause": "CALLED_ON_NULL_INPUT"
  },
  "schema": "default",
  "typeVariableConstraints": [
    {
      "comparableRequired": true,
      "name": "t",
      "nonDecimalNumericRequired": false,
      "orderableRequired": true,
      "variadicBound": ""
    }
  ],
  "variableArity": false
}
```

In this case:
- The functionHandle passed is an instance of NativeFunctionHandle.
- Let's suppose the signature that gets passed down is `native.default.max(bigint):bigint`, we use the passed down param type `bigint` to resolve the intermediate type `t` as `bigint`.

`max_by` function: 

```
{
  "aggregateMetadata": {
    "intermediateType": "row(bigint,array(boolean),array(v))",
    "isOrderSensitive": true
  },
  "docString": "native.default.max_by",
  "functionKind": "AGGREGATE",
  "outputType": "array(v)",
  "paramTypes": [
    "v",
    "boolean",
    "bigint"
  ],
  "routineCharacteristics": {
    "determinism": "DETERMINISTIC",
    "language": "CPP",
    "nullCallClause": "CALLED_ON_NULL_INPUT"
  },
  "schema": "default",
  "typeVariableConstraints": [
    {
      "comparableRequired": false,
      "name": "v",
      "nonDecimalNumericRequired": false,
      "orderableRequired": false,
      "variadicBound": ""
    }
  ],
  "variableArity": false
}
```

In this case:

- The functionHandle passed is an instance of NativeFunctionHandle.
- Let's suppose the signature that gets passed down is `native.default.max_by(integer, boolean, bigint):array(integer)`, we use the passed down param types `integer ` to resolve the intermediate type `"intermediateType": "row(bigint,array(boolean),array(v))` as `"intermediateType": "row(bigint,array(boolean),array(integer))`